salt/modules/ansiblegate.py

```diff
@@ -222,99 +58,202 @@ def _set_function(cmd_name, doc):
-        try:
-            docset = salt.utils.yaml.safe_load(docset)
-            if docset:
-                doc.update(docset)
-        except Exception as err:  # pylint: disable=broad-except
-            log.error("Error parsing doc section: %s", err)
+    ansible_doc_bin = salt.utils.path.which("ansible-doc")
+
+    proc = subprocess.run(
+        [ansible_doc_bin, "--json", "--type=module", module],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        check=True,
+        shell=False,
+        universal_newlines=True,
+    )
+    data = salt.utils.json.loads(proc.stdout)
+    doc = data[next(iter(data))]
     if not args:
-        if "description" in doc:
-            description = doc.get("description") or ""
-            del doc["description"]
-            ret["Description"] = description
-        ret[
-            'Available sections on module "{}"'.format(
-                module.__name__.replace("ansible.modules.", "")
-            )
-        ] = list(doc)
+        ret = doc["doc"]
+        for section in ("examples", "return", "metadata"):
+            section_data = doc.get(section)
+            if section_data:
+                ret[section] = section_data
     else:
+        ret = {}
         for arg in args:
             info = doc.get(arg)
             if info is not None:
                 ret[arg] = info
-
     return ret
 
 
-@depends("ansible")
 def list_(pattern=None):
     """
     Lists available modules.
-    :return:
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt * ansible.list
+        salt * ansible.list '*win*'  # To get all modules matching 'win' on it's name
+    """
+    if pattern is None:
+        module_list = set(__load__)
+        module_list.discard(set(__non_ansible_functions__))
+        return sorted(module_list)
+    return sorted(fnmatch.filter(__load__, pattern))
+
+
+def call(module, *args, **kwargs):
     """
-    return _resolver.get_modules_list(pattern=pattern)
+    Call an Ansible module by invoking it.
+
+    :param module: the name of the module.
+    :param args: Arguments to pass to the module
+    :param kwargs: keywords to pass to the module
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt * ansible.call ping data=foobar
+    """
+
+    module_args = []
+    for arg in args:
+        module_args.append(salt.utils.json.dumps(arg))
+
+    _kwargs = {}
+    for _kw in kwargs.get("__pub_arg", []):
+        if isinstance(_kw, dict):
+            _kwargs = _kw
+            break
+    else:
+        _kwargs = {k: v for (k, v) in kwargs.items() if not k.startswith("__pub")}
+
+    for key, value in _kwargs.items():
+        module_args.append("{}={}".format(key, salt.utils.json.dumps(value)))
+
+    with NamedTemporaryFile(mode="w") as inventory:
+
+        ansible_binary_path = salt.utils.path.which("ansible")
+        log.debug("Calling ansible module %r with arguments %r", module, module_args)
```

Comment:
Would it be possible for this to log any sensitive information, like passwords?

Reply:
If ansible allows passing passwords as arguments, yes.
Don't really know.

Take that log out?

Reply:
Yeah, or include: `log.debug("Calling ansible module %r ", module)` If that is still of use.

Reply:
:+1: